File: src/CurrencyPal/Services/RateService.cs
using CurrencyPal.Dto;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;

namespace CurrencyPal.Services
{
    public class RateService : IRateService
    {
        private readonly HttpClient _Client = new HttpClient();
        private readonly string _CurrencyLayerAppId = Environment.GetEnvironmentVariable("CURRENCYLAYER_APPID");
        private readonly string[] _Tickers = { "USDZAR", "GBPZAR", "ZARMUR", "BTCZAR", "BTCUSD" };

        public async Task<RatesDto> GetExchangeRates()
        {
            var bitcoin = await GetBitcoin();
            var sourceRates = await GetRates();
            
            var dto = new RatesDto();
            foreach (var ticker in _Tickers)
            {
                double rate = ticker == "BTCZAR" ? bitcoin : GetRate(ticker, sourceRates);
                
                dto.Rates.Add(new RateDto
                {
                    Rate = rate,
                    Ticker = ticker
                });
            }

            return dto;
        }

        private async Task<double> GetBitcoin()
        {
            var response = await _Client.GetStringAsync($"https://api.mybitx.com/api/1/ticker?pair=XBTZAR");
            var rate = JObject.Parse(response)["last_trade"].ToObject<double>();
            return rate;
        }

        private async Task<Dictionary<string, double>> GetRates()
        {
            var ratesResponse = await _Client.GetStringAsync($"https://openexchangerates.org/api/latest.json?app_id={_CurrencyLayerAppId}");
            var sourceRates = JObject.Parse(ratesResponse)["rates"].ToObject<Dictionary<string, double>>();
            return sourceRates;
        }

        private static double GetRate(string ticker, Dictionary<string, double> rates)
        {
            var source = ticker.Substring(0, 3);
            var dest = ticker.Substring(3, 3);

            var sourceRate = rates[source];
            var destRate = rates[dest];
            var crossRate = destRate / sourceRate;
            return crossRate;
        }
    }
}

File: services/rateService.js
const fetch = require("node-fetch");

const tickers = ["USDZAR", "GBPZAR", "ZARMUR", "BTCZAR", "BTCUSD"];

var cacheExpiryTime = Date.now();
var cachedRates;

var getRates = function() {

    // if the cached rates are still valid
    if (Date.now() < cacheExpiryTime) {
        return Promise.resolve(cachedRates)
    }

    // get rates
    var bitcoinPromise = getBitcoin();
    var otherRatesPromise = getOtherRates();

    return Promise.all([bitcoinPromise, otherRatesPromise])
        .then(results => {
            var bitcoin = results[0];
            var otherRates = results[1];
            var rates = tickers.map(ticker => ticker === "BTCZAR" ? bitcoin : otherRates.find(r => r.ticker === ticker));
            
            updateTime = Date.now();
            cacheExpiryTime = updateTime + 1000 * 60 * 60 // 1 hour from now
            cachedRates = {
                rates: rates,
                updateTime: updateTime
            };

            return cachedRates;
        })
}

var getBitcoin = function() {
    return fetch('https://api.mybitx.com/api/1/ticker?pair=XBTZAR')
        .then(res => res.json())
        .then(res => {
            var r = {ticker: "BTCZAR", rate: parseFloat(res.last_trade)};
            return r;
        });
}

var getOtherRates = function() {
    var appId = process.env.OPENEXCHANGERATES_APPID;
    return fetch("https://openexchangerates.org/api/latest.json?app_id=" + appId)
        .then(res => res.json())
        .then(res => {
            var source = res.rates;
            var rates = tickers.map(x => convertRate(x, source));
            return rates;
        });
}

var convertRate = function(ticker, rates){
    var source = ticker.substring(0, 3);
    var dest = ticker.substring(3, 6);

    var sourceRate = rates[source];
    var destRate = rates[dest];
    var crossRate = destRate / sourceRate;
    return {ticker: ticker, rate: crossRate};
}

module.exports = getRates;